File: LeetCode/DP/Rust/Leetcode828-count-unique-characters-of-all-substrings-of-a-given-string.rs
impl Solution {
    pub fn unique_letter_string(s: String) -> i32 {
        let n = s.len();
        let (mut l,mut r,mut last) = (vec![0;n],vec![0;n],vec![-1;26]);
        let sb = s.as_bytes();
        for i in 0..n {
            l[i] = last[(sb[i] - b'A') as usize];
            last[(sb[i] - b'A') as usize] = i as i32;
        }
        last = vec![n as i32;26];
        for i in (0..n).rev() {
            r[i] = last[(sb[i] - b'A') as usize];
            last[(sb[i] - b'A') as usize] = i as i32;
        }
        let mut ans = 0;
        for i in 0..n {
            ans += (i as i32 - l[i]) * (r[i] - i as i32)
        }
        ans
    }
}

File: LeetCode/DP/Python/Leetcode828-count-unique-characters-of-all-substrings-of-a-given-string.py
class Solution:
    def uniqueLetterString(self, s: str) -> int:
        n = len(s)
        l, r, last = [0] * n, [0] * n, [-1] * 26
        for i in range(n):
            l[i] = last[ord(s[i]) - ord('A')]
            last[ord(s[i]) - ord('A')] = i
        last = [n] * 26
        for i in range(n - 1, -1, -1):
            r[i] = last[ord(s[i]) - ord('A')]
            last[ord(s[i]) - ord('A')] = i
        ans = 0
        for i in range(n):
            ans += (i - l[i]) * (r[i] - i)
        return ans
